algebra-core/src/fields/macros.rs

```diff
@@ -1,3 +1,535 @@
+                    // Guajardo Kumar Paar Pelzl
+                    // Efficient Software-Implementation of Finite Fields with Applications to
+                    // Cryptography
+                    // Algorithm 16 (BEA for Inversion in Fp)
+
+                    let one = BigInteger::from(1);
+
+                    let mut u = self.0;
+                    let mut v = P::MODULUS;
+                    let mut b = $Fp::<P>(P::R2, PhantomData); // Avoids unnecessary reduction step.
+                    let mut c = Self::zero();
+
+                    while u != one && v != one {
+                        while u.is_even() {
+                            u.div2();
+
+                            if b.0.is_even() {
+                                b.0.div2();
+                            } else {
+                                b.0.add_nocarry(&P::MODULUS);
+                                b.0.div2();
+                            }
+                        }
+
+                        while v.is_even() {
+                            v.div2();
+
+                            if c.0.is_even() {
+                                c.0.div2();
+                            } else {
+                                c.0.add_nocarry(&P::MODULUS);
+                                c.0.div2();
+                            }
+                        }
+
+                        if v < u {
+                            u.sub_noborrow(&v);
+                            b.sub_assign(&c);
+                        } else {
+                            v.sub_noborrow(&u);
+                            c.sub_assign(&b);
+                        }
+                    }
+
+                    if u == one {
+                        Some(b)
+                    } else {
+                        Some(c)
+                    }
+                }
+            }
+
+            fn inverse_in_place(&mut self) -> Option<&mut Self> {
+                if let Some(inverse) = self.inverse() {
+                    *self = inverse;
+                    Some(self)
+                } else {
+                    None
+                }
+            }
+
+            #[inline]
+            fn frobenius_map(&mut self, _: usize) {
+                // No-op: No effect in a prime field.
+            }
+        }
+
+        impl<P: $FpParameters> PrimeField for $Fp<P> {
+            type Params = P;
+            type BigInt = BigInteger;
+
+            #[inline]
+            fn from_repr(r: BigInteger) -> Self {
+                let mut r = $Fp(r, PhantomData);
+                if r.is_valid() {
+                    r.mul_assign(&$Fp(P::R2, PhantomData));
+                    r
+                } else {
+                    Self::zero()
+                }
+            }
+
+            impl_field_into_repr!($limbs);
+
+            #[inline]
+            fn from_random_bytes(bytes: &[u8]) -> Option<Self> {
+                let mut result_bytes = vec![0u8; (Self::zero().0).0.len() * 8];
```

Comment:
We can also avoid allocations:

(This is code that I wrote, but I noticed it while reviewing the bug)
```suggestion
                let mut result_bytes = [0u8; $limbs * 8];
```

Reply:
Great, I was looking for this bug!